In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
df = pd.read_excel("C:/Users/Hp/Downloads/Hydroponics_Analytics_Dataset_10_Days.xlsx")

In [19]:
#Dataset Overview

In [5]:
df.head(10)

,Timestamp,pH,TDS_ppm,Air_Temperature_C,Water_Temperature_C,Humidity_pct,Water_Level_pct,Water_Pump,Nutrient_Pump,Humidifier,Exhaust_Fan,Alert
0,2026-03-01 00:00,6.16,709,26.1,23.9,62,100,OFF,ON,OFF,OFF,Low TDS
1,2026-03-01 01:00,6.26,704,26.2,24.3,65,98,OFF,ON,OFF,OFF,Low TDS
2,2026-03-01 02:00,6.15,718,25.9,23.6,65,96,OFF,OFF,OFF,OFF,Normal
3,2026-03-01 03:00,6.22,715,25.5,23.3,62,94,OFF,OFF,OFF,OFF,Normal
4,2026-03-01 04:00,6.26,707,25.9,23.8,64,92,OFF,ON,OFF,OFF,Low TDS
5,2026-03-01 05:00,6.27,712,25.8,23.6,61,90,OFF,ON,ON,OFF,Low TDS
6,2026-03-01 06:00,6.26,710,26.6,24.4,61,88,OFF,ON,ON,OFF,Low TDS
7,2026-03-01 07:00,6.28,714,27.2,25.3,61,86,OFF,ON,ON,OFF,Low TDS
8,2026-03-01 08:00,6.22,713,27.1,25.2,65,84,OFF,ON,OFF,OFF,Low TDS
9,2026-03-01 09:00,6.23,711,26.7,24.4,61,82,OFF,ON,ON,OFF,Low TDS


In [4]:
df.tail()

,Timestamp,pH,TDS_ppm,Air_Temperature_C,Water_Temperature_C,Humidity_pct,Water_Level_pct,Water_Pump,Nutrient_Pump,Humidifier,Exhaust_Fan,Alert
235,2026-03-10 19:00,6.37,722,27.7,25.5,64,53,OFF,OFF,OFF,OFF,Normal
236,2026-03-10 20:00,6.33,722,27.7,25.8,66,51,OFF,OFF,OFF,OFF,Normal
237,2026-03-10 21:00,6.42,733,27.9,25.9,66,49,OFF,OFF,OFF,OFF,Normal
238,2026-03-10 22:00,6.38,723,27.7,25.7,66,47,OFF,OFF,OFF,OFF,Normal
239,2026-03-10 23:00,6.40,729,28.2,26.2,64,45,OFF,OFF,OFF,OFF,Normal


In [6]:
df.shape

(240, 12)

In [7]:
df.columns

Index(['Timestamp', 'pH', 'TDS_ppm', 'Air_Temperature_C',
       'Water_Temperature_C', 'Humidity_pct', 'Water_Level_pct', 'Water_Pump',
       'Nutrient_Pump', 'Humidifier', 'Exhaust_Fan', 'Alert'],
      dtype='str')

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Timestamp            240 non-null    str    
 1   pH                   240 non-null    float64
 2   TDS_ppm              240 non-null    int64  
 3   Air_Temperature_C    240 non-null    float64
 4   Water_Temperature_C  240 non-null    float64
 5   Humidity_pct         240 non-null    int64  
 6   Water_Level_pct      240 non-null    int64  
 7   Water_Pump           240 non-null    str    
 8   Nutrient_Pump        240 non-null    str    
 9   Humidifier           240 non-null    str    
 10  Exhaust_Fan          240 non-null    str    
 11  Alert                240 non-null    str    
dtypes: float64(3), int64(3), str(6)
memory usage: 22.6 KB


In [20]:
#Data Quality

In [9]:
df.isnull().sum()

Timestamp              0
pH                     0
TDS_ppm                0
Air_Temperature_C      0
Water_Temperature_C    0
Humidity_pct           0
Water_Level_pct        0
Water_Pump             0
Nutrient_Pump          0
Humidifier             0
Exhaust_Fan            0
Alert                  0
dtype: int64

In [10]:
df.duplicated().sum()

np.int64(0)

In [11]:
df.dtypes

Timestamp                  str
pH                     float64
TDS_ppm                  int64
Air_Temperature_C      float64
Water_Temperature_C    float64
Humidity_pct             int64
Water_Level_pct          int64
Water_Pump                 str
Nutrient_Pump              str
Humidifier                 str
Exhaust_Fan                str
Alert                      str
dtype: object

In [12]:
df.describe()

,pH,TDS_ppm,Air_Temperature_C,Water_Temperature_C,Humidity_pct,Water_Level_pct
count,240.000000,240.000000,240.000000,240.000000,240.000000,240.000000
mean,6.314458,721.545833,30.072917,27.858750,63.850000,70.812500
std,0.084416,7.545537,3.050368,2.722771,2.583015,14.781047
min,6.120000,702.000000,25.500000,23.300000,58.000000,36.000000
25%,6.250000,716.000000,27.600000,25.500000,62.000000,59.000000
50%,6.320000,722.000000,29.200000,27.200000,64.000000,70.500000
75%,6.370000,727.000000,32.800000,30.650000,66.000000,83.000000
max,6.470000,738.000000,37.500000,32.000000,69.000000,100.000000


In [21]:
# Summary Statistics

In [13]:
df.describe(include="object")

C:\Users\Hp\AppData\Local\Temp\ipykernel_8160\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,Timestamp,Water_Pump,Nutrient_Pump,Humidifier,Exhaust_Fan,Alert
count,240,240,240,240,240,240
unique,240,2,2,2,2,6
top,2026-03-01 00:00,OFF,OFF,OFF,OFF,Normal
freq,1,237,193,192,168,129


In [22]:
# Unique Values

In [14]:
df["Alert"].value_counts()

Alert
Normal                                        129
High Temperature                               62
Low TDS                                        39
Low TDS; High Temperature                       7
Low Water Level; High Temperature               2
Low Water Level; Low TDS; High Temperature      1
Name: count, dtype: int64

In [15]:
df["Water_Pump"].value_counts()

Water_Pump
OFF    237
ON       3
Name: count, dtype: int64

In [16]:
df["Nutrient_Pump"].value_counts()

Nutrient_Pump
OFF    193
ON      47
Name: count, dtype: int64

In [17]:
df["Humidifier"].value_counts()

Humidifier
OFF    192
ON      48
Name: count, dtype: int64

In [18]:
df["Exhaust_Fan"].value_counts()

Exhaust_Fan
OFF    168
ON      72
Name: count, dtype: int64

In [23]:
# Check Data Range

In [24]:
df.min(numeric_only=True)

pH                       6.12
TDS_ppm                702.00
Air_Temperature_C       25.50
Water_Temperature_C     23.30
Humidity_pct            58.00
Water_Level_pct         36.00
dtype: float64

In [25]:
df.max(numeric_only=True)

pH                       6.47
TDS_ppm                738.00
Air_Temperature_C       37.50
Water_Temperature_C     32.00
Humidity_pct            69.00
Water_Level_pct        100.00
dtype: float64

In [26]:
# Convert Timestamp

In [27]:
df["Timestamp"] = pd.to_datetime(df["Timestamp"])

In [28]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 240 entries, 0 to 239
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Timestamp            240 non-null    datetime64[us]
 1   pH                   240 non-null    float64       
 2   TDS_ppm              240 non-null    int64         
 3   Air_Temperature_C    240 non-null    float64       
 4   Water_Temperature_C  240 non-null    float64       
 5   Humidity_pct         240 non-null    int64         
 6   Water_Level_pct      240 non-null    int64         
 7   Water_Pump           240 non-null    str           
 8   Nutrient_Pump        240 non-null    str           
 9   Humidifier           240 non-null    str           
 10  Exhaust_Fan          240 non-null    str           
 11  Alert                240 non-null    str           
dtypes: datetime64[us](1), float64(3), int64(3), str(5)
memory usage: 22.6 KB


## Initial Observations

- Dataset contains 240 hourly observations collected over 10 days.
- No missing values were detected.
- No duplicate records were found.
- pH values remained within the acceptable range.
- Temperature increased significantly during Days 4–6.
- Water level gradually decreased before pump activation.
- Nutrient pump was activated only when TDS dropped below the threshold.